src/ResourceStore.php

```diff
@@ -0,0 +1,237 @@
+
+    public function find($type, $id, $query_params = [])
+    {
+        if (!$this->hasResource($type, $id)) {
+            $result = $this->http_client->request(
+                'GET',
+                $this->getResourceAddress($type, $id),
+                ['query' => $query_params]
+            );
+
+            $body = json_decode($result->getBody(), true);
+
+            $resource = new Resource($this);
+            $resource->decode($body['data']);
+
+            $this->setResource($type, $id, $resource);
+        }
+
+        return $this->getResource($type, $id);
+    }
+
+    // }}}
+    // {{{ public function save()
+
+    public function save(Resource $resource)
+    {
+        $result = $this->http_client->request(
+            'PATCH',
+            $this->getResourceAddress(
+                $resource->getType(),
+                $resource->getId()
+            ),
+            ['body' => $resource->encode()]
+        );
+
+        $body = json_decode($result->getBody(), true);
+
+        $resource = new Resource($this);
+        $resource->decode($body['data']);
+
+        // Replace the old resource with a new one
+        $this->setResource(
+            $resource->getType(),
+            $resource->getId(),
+            $resource
+        );
+
+        return $this->getResource(
+            $resource->getType(),
+            $resource->getId()
+        );
+    }
+
+    // }}}
+    // {{{ public function create()
+
+    public function create(Resource $resource)
+    {
+        $result = $this->http_client->request(
+            'POST',
+            $this->getResourceAddress($resource->getType()),
+            ['body' => $resource->encode()]
+        );
+
+        $body = json_decode($result->getBody(), true);
+
+        $resource = new Resource($this);
+        $resource->decode($body['data']);
+
+        // Replace the old resource with a new one
+        $this->setResource(
+            $resource->getType(),
+            $resource->getId(),
+            $resource
+        );
+
+        return $this->getResource(
+            $resource->getType(),
+            $resource->getId()
+        );
+    }
+
+    // }}}
+    // {{{ protected function hasResource()
+
+    protected function hasResource($type, $id)
+    {
+        return isset($this->resources[$type][$id]);
+    }
+
+    // }}}
+    // {{{ protected function getResource()
+
+    protected function getResource($type, $id)
+    {
+        return $this->resources[$type][$id];
+    }
+
+    // }}}
+    // {{{ protected function setResource()
+
+    protected function setResource($type, $id, Resource $resource)
+    {
+        return $this->resources[$type][$id] = $resource;
+    }
+
+    // }}}
+    // {{{ protected function getResourceAddress()
+
+    protected function getResourceAddress($type, $id = null)
+    {
+        return ($id != '') ? $type . '/' . $id : $type;
+    }
+
+    // }}}
+    // {{{ public function __sleep()
+
+    public function __sleep()
```

Comment:
We've avoided `__sleep()/__wakeup()` in favor of the Serializable interface due to memory corruption bugs. Not sure if this is still required but I lean toward implementing the interface.